File: go/mechanisms/evm/exact/facilitator/erc20_approval.go
package facilitator

import (
	"encoding/hex"
	"fmt"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"

	"github.com/x402-foundation/x402/go/extensions/erc20approvalgassponsor"
	"github.com/x402-foundation/x402/go/mechanisms/evm"
)

// approveSelector is the 4-byte function selector for ERC-20 approve(address,uint256).
// keccak256("approve(address,uint256)") = 0x095ea7b3...
var approveSelector = []byte{0x09, 0x5e, 0xa7, 0xb3}

// ValidateErc20ApprovalForPayment validates the ERC-20 approval extension data.
// Returns ("", "") if valid, or (reason, message) on failure where reason is the
// error constant and message is a human-readable contextual description.
func ValidateErc20ApprovalForPayment(info *erc20approvalgassponsor.Info, payer, tokenAddress string) (reason, message string) {
	if !erc20approvalgassponsor.ValidateInfo(info) {
		return ErrErc20ApprovalInvalidFormat, "ERC-20 approval extension info failed format validation"
	}

	if !strings.EqualFold(info.From, payer) {
		return ErrErc20ApprovalFromMismatch, fmt.Sprintf("expected from=%s, got %s", payer, info.From)
	}

	if !strings.EqualFold(info.Asset, tokenAddress) {
		return ErrErc20ApprovalAssetMismatch, fmt.Sprintf("expected asset=%s, got %s", tokenAddress, info.Asset)
	}

	if !strings.EqualFold(info.Spender, evm.PERMIT2Address) {
		return ErrErc20ApprovalWrongSpender, fmt.Sprintf("expected spender=%s, got %s", evm.PERMIT2Address, info.Spender)
	}

	txHex := strings.TrimPrefix(info.SignedTransaction, "0x")
	txBytes, err := hex.DecodeString(txHex)
	if err != nil {
		return ErrErc20ApprovalTxParseFailed, "failed to hex-decode signed transaction"
	}

	tx := new(types.Transaction)
	if err := tx.UnmarshalBinary(txBytes); err != nil {
		return ErrErc20ApprovalTxParseFailed, "failed to RLP-decode signed transaction"
	}

	txTo := ""
	if tx.To() != nil {
		txTo = tx.To().Hex()
	}
	if tx.To() == nil || !strings.EqualFold(txTo, tokenAddress) {
		return ErrErc20ApprovalWrongTarget, fmt.Sprintf("transaction targets %s, expected %s", txTo, tokenAddress)
	}

	data := tx.Data()
	if len(data) < 4 {
		return ErrErc20ApprovalWrongSelector, "transaction calldata too short for approve() selector"
	}
	for i, b := range approveSelector {
		if data[i] != b {
			return ErrErc20ApprovalWrongSelector, "transaction calldata does not start with approve() selector 0x095ea7b3"
		}
	}

	if len(data) < 36 {
		return ErrErc20ApprovalWrongCalldata, "transaction calldata too short to contain spender parameter"
	}
	calldataSpender := common.BytesToAddress(data[4:36])
	if !strings.EqualFold(calldataSpender.Hex(), evm.PERMIT2Address) {
		return ErrErc20ApprovalWrongCalldata, fmt.Sprintf("approve() spender is %s, expected Permit2 %s", calldataSpender.Hex(), evm.PERMIT2Address)
	}

	chainID := tx.ChainId()
	signer := types.LatestSignerForChainID(chainID)
	from, err := types.Sender(signer, tx)
	if err != nil {
		return ErrErc20ApprovalInvalidSig, "failed to recover signer from the signed transaction"
	}
	if !strings.EqualFold(from.Hex(), payer) {
		return ErrErc20ApprovalSignerMismatch, fmt.Sprintf("transaction signed by %s, expected payer %s", from.Hex(), payer)
	}

	return "", ""
}

File: typescript/packages/mechanisms/evm/src/shared/erc20approval.ts
import {
  getAddress,
  parseTransaction,
  decodeFunctionData,
  recoverTransactionAddress,
  type TransactionSerialized,
} from "viem";
import type { VerifyResponse } from "@x402/core/types";
import {
  validateErc20ApprovalGasSponsoringInfo,
  type Erc20ApprovalGasSponsoringInfo,
} from "../exact/extensions";
import { PERMIT2_ADDRESS, erc20ApproveAbi } from "../constants";
import {
  ErrErc20ApprovalInvalidFormat,
  ErrErc20ApprovalFromMismatch,
  ErrErc20ApprovalAssetMismatch,
  ErrErc20ApprovalSpenderNotPermit2,
  ErrErc20ApprovalTxWrongTarget,
  ErrErc20ApprovalTxWrongSelector,
  ErrErc20ApprovalTxWrongSpender,
  ErrErc20ApprovalTxInvalidCalldata,
  ErrErc20ApprovalTxSignerMismatch,
  ErrErc20ApprovalTxInvalidSignature,
  ErrErc20ApprovalTxParseFailed,
} from "../exact/facilitator/errors";

/** The approve(address,uint256) function selector */
const APPROVE_SELECTOR = "0x095ea7b3";

/**
 * Validates ERC-20 approval extension data for a Permit2 payment.
 *
 * Performs comprehensive validation:
 * - Format validation via validateErc20ApprovalGasSponsoringInfo (JSON Schema)
 * - `from` matches payer
 * - `asset` matches token
 * - `spender` is PERMIT2_ADDRESS
 * - Transaction `to` matches token address
 * - Transaction calldata is a valid approve(PERMIT2_ADDRESS, ...) call
 * - Recovered transaction signer matches `from`
 *
 * @param info - The ERC-20 approval gas sponsoring info
 * @param payer - The expected payer address
 * @param tokenAddress - The expected token address
 * @returns Validation result with invalidReason and invalidMessage on failure
 */
export async function validateErc20ApprovalForPayment(
  info: Erc20ApprovalGasSponsoringInfo,
  payer: `0x${string}`,
  tokenAddress: `0x${string}`,
): Promise<Pick<VerifyResponse, "isValid" | "invalidReason" | "invalidMessage">> {
  if (!validateErc20ApprovalGasSponsoringInfo(info)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalInvalidFormat,
      invalidMessage: "ERC-20 approval extension info failed schema validation",
    };
  }

  if (getAddress(info.from) !== getAddress(payer)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalFromMismatch,
      invalidMessage: `Expected from=${payer}, got ${info.from}`,
    };
  }

  if (getAddress(info.asset) !== tokenAddress) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalAssetMismatch,
      invalidMessage: `Expected asset=${tokenAddress}, got ${info.asset}`,
    };
  }

  if (getAddress(info.spender) !== getAddress(PERMIT2_ADDRESS)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalSpenderNotPermit2,
      invalidMessage: `Expected spender=${PERMIT2_ADDRESS}, got ${info.spender}`,
    };
  }

  try {
    const serializedTx = info.signedTransaction as TransactionSerialized;
    const tx = parseTransaction(serializedTx);

    if (!tx.to || getAddress(tx.to) !== tokenAddress) {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxWrongTarget,
        invalidMessage: `Transaction targets ${tx.to ?? "null"}, expected ${tokenAddress}`,
      };
    }

    const data = tx.data ?? "0x";
    if (!data.startsWith(APPROVE_SELECTOR)) {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxWrongSelector,
        invalidMessage: `Transaction calldata does not start with approve() selector ${APPROVE_SELECTOR}`,
      };
    }

    try {
      const decoded = decodeFunctionData({
        abi: erc20ApproveAbi,
        data: data as `0x${string}`,
      });
      const calldataSpender = getAddress(decoded.args[0] as `0x${string}`);
      if (calldataSpender !== getAddress(PERMIT2_ADDRESS)) {
        return {
          isValid: false,
          invalidReason: ErrErc20ApprovalTxWrongSpender,
          invalidMessage: `approve() spender is ${calldataSpender}, expected Permit2 ${PERMIT2_ADDRESS}`,
        };
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxInvalidCalldata,
        invalidMessage: "Failed to decode approve() calldata from the signed transaction",
      };
    }

    try {
      const recoveredAddress = await recoverTransactionAddress({
        serializedTransaction: serializedTx,
      });
      if (getAddress(recoveredAddress) !== getAddress(payer)) {
        return {
          isValid: false,
          invalidReason: ErrErc20ApprovalTxSignerMismatch,
          invalidMessage: `Transaction signed by ${recoveredAddress}, expected payer ${payer}`,
        };
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxInvalidSignature,
        invalidMessage: "Failed to recover signer from the signed transaction",
      };
    }
  } catch {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalTxParseFailed,
      invalidMessage: "Failed to parse the signed transaction",
    };
  }

  return { isValid: true };
}
